### DISCUSSION TITLE
numpy array of arrays?

### DISCUSSION THREAD

**Discussion by cary-ilm** [pybind/pybind11#5128] [Q&A]
With numpy, I can create a 2D array of arrays, where each entry has a unique length:

```
array_of_arrays = np.empty((3, 3), dtype=object)
for i in range(3):
    for j in range(3):
        array_of_arrays[i, j] = np.array([k for k in range(i*3+j)])
```

Is it possible to create such a thing in pybind11? When I try to create ``py::array_t<py::array_t<int>>``, I get:

` static assertion failed: Attempt to use a non-POD or unimplemented POD type as a numpy dtype
`

which makes it appear that ``py::array`` can't handle non-POD objects like ``py::array`` as a type, unless there's a way to specify it that I'm missing. The documentation also appears to imply this, but I'd like to confirm I'm understanding the limitation before resorting to an alternative design.

The specific application is deep image data: a 2D array of pixels, where each pixel is an array of samples, in case there's a better way of achieving that. A numpy array of lists would suffice, but presumably that has the same problem. The alternative of a list of lists of lists/arrays is less desirable.

**Comment by dkoo2** [comment]
What worked for me is a signature like `int f(::pybind11::array my_array)`, where `my_array` can be an np.array of np.arrays where the inner arrays are different length.

Then you can access the inner arrays like

```
::pybind11::buffer_info buffer_info = my_array.request();
auto *ptr = static_cast<::pybind11::array*>(buffer_info.ptr);
```

If you just need a 2D np array then your type should be `py_array_t<int>` instead
